Jenkinsfile.macos

```diff
@@ -1,175 +1,283 @@
+def utils
+
 pipeline {
 
   agent { label 'macos-v1.4-arm64' }
 
   options {
+    // timeout regardless of activity
+    timeout(activity: false, time: 120)
     timestamps()
     disableConcurrentBuilds()
     buildDiscarder(logRotator(numToKeepStr: '100'))
   }
 
   parameters {
-      string(name: 'RSTUDIO_VERSION_MAJOR',  defaultValue: '1', description: 'RStudio Major Version')
-      string(name: 'RSTUDIO_VERSION_MINOR',  defaultValue: '1', description: 'RStudio Minor Version')
-      string(name: 'RSTUDIO_VERSION_PATCH',  defaultValue: '999', description: 'RStudio Patch Version')
-      string(name: 'RSTUDIO_VERSION_SUFFIX', defaultValue: '+9999', description: 'RStudio Pro Suffix Version')
-      string(name: 'GIT_REVISION', defaultValue: 'main', description: 'Git revision to build')
-      string(name: 'BRANCH_NAME', defaultValue: 'origin/main', description: 'Branch name to build')
       string(name: 'SLACK_CHANNEL', defaultValue: '#ide-builds', description: 'Slack channel to publish build message.')
+      string(name: 'RSTUDIO_VERSION_PATCH',  defaultValue: '999', description: 'RStudio Patch Version')
+      booleanParam(name: 'DAILY', defaultValue: true, description: 'Runs daily build if true')
+      booleanParam(name: 'PUBLISH', defaultValue: true, description: 'Runs publish stage if true')
+      gitParameter defaultValue: "${env.GIT_BRANCH}", name: 'COMMIT_HASH', type: 'PT_REVISION'
   }
 
   environment {
     PATH = "$HOME/opt/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin"
     PACKAGE_OS = 'macOS'
+    RSTUDIO_VERSION = ""
+    RSTUDIO_VERSION_MAJOR = 0
+    RSTUDIO_VERSION_MINOR = 0
+    RSTUDIO_VERSION_PATCH = 0
+    RSTUDIO_VERSION_SUFFIX = 0
+    IS_PRO = false
   }
-
   stages {
 
-    stage('dependencies') {
-
-      environment {
-        // boost won't compile without the brew version of openssl.
-        // only add it to the dep resolve step though, or the ide build will compile against the wrong openssl
-        PATH = '/usr/local/opt/openssl/bin:/usr/local/opt/openssl/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin'
-      }
-
+    stage('Load Utils') {
       steps {
         script {
-          currentBuild.displayName = "${params.RSTUDIO_VERSION_MAJOR}.${params.RSTUDIO_VERSION_MINOR}.${params.RSTUDIO_VERSION_PATCH}${params.RSTUDIO_VERSION_SUFFIX}"
-        }
-        withCredentials([usernameColonPassword(credentialsId: 'github-rstudio-jenkins', variable: "GITHUB_LOGIN")]) {
-          sh 'cd dependencies/osx && RSTUDIO_GITHUB_LOGIN=$GITHUB_LOGIN ./install-dependencies-osx && cd ../..'
+          sh 'printenv'
+          sh "echo 'Loading utils from ${env.WORKSPACE}/utils.groovy'"
+          utils = load "${env.WORKSPACE}/utils.groovy"
         }
       }
-
     }
 
-    stage('build') {
+    stage ("Checkout") {
+      when {
+        expression { COMMIT_HASH != null}
```

Comment:
So I went back and did some testing here, cause I could never get it to work in my experimental pipeline.

It seems to me that we are unable to use environment variables inside of the top `parameters` block. I would just end up with a `null` value and the checkout would fail.

I tried comparing against `null` with
```
when {
   expression { return COMMIT_HASH != null }
}
```

Both passing in `defaultValue: 'none'` and `defaultValue: null` would result in:
```
[2023-01-19T22:09:51.944Z] Commit_hash value: null
```
Which meant that in the "Checkout" stage, the `when` block check did not work when doing a null comparison and it ran the stage anyways.

The only solution I could get to work is passing in `defaultValue: 'none'` and doing:
```
when {
  expression { params.COMMIT_HASH != 'none' }
}
```
With that, it would correctly skip the stage. So that works, but I'm definitely open to suggestions for a better approach.